app/src/main/java/com/samco/trackandgraph/ui/CustomiseCardDialog.kt

```diff
@@ -0,0 +1,170 @@
+
+            listenToViewModelState()
+            binding.root
+        } ?: throw IllegalStateException("Activity cannot be null")
+    }
+
+    private fun getOnPeriodSelectedListener(): AdapterView.OnItemSelectedListener {
+        return object : AdapterView.OnItemSelectedListener {
+            override fun onNothingSelected(p0: AdapterView<*>?) {}
+            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, position: Int, p3: Long) {
+                viewModel.setPeriodTo(featurePeriodList[position])
+            }
+        }
+    }
+    private fun getOnMethodSelectedListener(): AdapterView.OnItemSelectedListener {
+        return object : AdapterView.OnItemSelectedListener {
+            override fun onNothingSelected(p0: AdapterView<*>?) {}
+            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, position: Int, p3: Long) {
+                viewModel.setMethodTo(featureMethodList[position])
+            }
+        }
+    }
+    override fun onResume() {
+        super.onResume()
+        dialog?.window?.setLayout(
+            ViewGroup.LayoutParams.MATCH_PARENT,
+            ViewGroup.LayoutParams.WRAP_CONTENT
+        )
+    }
+
+    private fun listenToViewModelState() {
+        viewModel.state.observe(viewLifecycleOwner, Observer { state ->
+            when (state) {
+                CustomiseCardDialogState.INITIALIZING -> { setInitialViewState() }
+                CustomiseCardDialogState.WAITING -> { onViewModelReady() }
+                else -> {}
+            }
+        })
+    }
+
+    private fun setInitialViewState() {}
+    private fun onViewModelReady() {
+        binding.featureShowCountPeriodSpinner.setSelection(featurePeriodList.indexOf(viewModel.featurePeriod.value!!))
+        binding.featureShowCountMethodSpinner.setSelection(featureMethodList.indexOf(viewModel.featureMethod.value!!))
+        binding.featureShowCountPeriodSpinner.onItemSelectedListener = getOnPeriodSelectedListener()
+        binding.featureShowCountMethodSpinner.onItemSelectedListener = getOnMethodSelectedListener()
+        binding.closeButton.setOnClickListener { dismiss() }
+    }
+
+}
+
+enum class CustomiseCardDialogState { INITIALIZING, WAITING, UPDATING }  // UPDATING then WAITING again
+
+class CustomiseCardDialogViewModel : ViewModel() {
+    private var updateJob = Job()
+    private val ioScope = CoroutineScope(Dispatchers.IO + updateJob)
+    private var database: TrackAndGraphDatabase? = null
+    private var dao: TrackAndGraphDatabaseDao? = null
+
+    val state: LiveData<CustomiseCardDialogState> get() { return _state }
+    private val _state = MutableLiveData<CustomiseCardDialogState>(CustomiseCardDialogState.INITIALIZING)
+
+    val featurePeriod = MutableLiveData(FeatureShowCountPeriod.ALL)
+    val featureMethod = MutableLiveData(FeatureShowCountMethod.COUNT_ENTRIES)
+
+    private lateinit var feature: Feature
+
+    fun init(activity: Activity, id: Long) {
```

Comment:
I know I've done this everywhere but really you should pass the database dao in rather than the activity. A view model shouldn't really know about it's view.